cimbar/cimbar.py

```diff
@@ -135,31 +259,123 @@ def update(c, r, g, b):
+
+
+def _decode_symbols(ct, img):
     cell_pos, num_edge_cells = cell_positions(conf.CELL_SPACING_X, conf.CELL_SPACING_Y, conf.CELL_DIM_X,
                                               conf.CELL_DIM_Y, conf.CELLS_OFFSET, conf.MARKER_SIZE_X, conf.MARKER_SIZE_Y)
     finder = AdjacentCellFinder(cell_pos, num_edge_cells, conf.CELL_DIM_X, conf.MARKER_SIZE_X)
     decode_order = FloodDecodeOrder(cell_pos, finder)
+    print('beginning decode symbols pass...')
     for i, (x, y), drift in decode_order:
-        best_bits, best_dx, best_dy, best_distance = _decode_cell(ct, img, color_img, x, y, drift)
+        best_bits, best_cell, best_dx, best_dy, best_distance = _decode_cell(ct, img, x, y, drift)
         decode_order.update(best_dx, best_dy, best_distance)
-        yield i, best_bits
+        yield i, best_bits, best_cell
+
+
+def _calc_ccm(ct, color_lookups, cc_setting, state_info):
+    splits = 2 if cc_setting in (6, 7) else 0
+    if cc_setting in (3, 4, 5):
+        possible = possible_colors(ct.dark, BITS_PER_COLOR)
+        if len(color_lookups[0]) < len(possible):
+            raise Exception("kabooski")
+        exp = [color for i,color in enumerate(possible) if i in color_lookups[0]] + [(255,255,255)]
+        exp = numpy.array(exp)
+        white = state_info['white']
+        observed = numpy.array([v for k,v in sorted(color_lookups[0].items())] + [white])
+        from colour.characterisation.correction import matrix_colour_correction_Cheung2004
+        der = matrix_colour_correction_Cheung2004(observed, exp)
+
+        # not sure which of this would be better...
+        if ct.ccm is None or cc_setting == 4:
+            ct.ccm = der
+        else:  # cc_setting == 3,5
+            ct.ccm = der.dot(ct.ccm)
+
+    if splits:  # 6,7
+        from colour.characterisation.correction import matrix_colour_correction_Cheung2004
+        exp = numpy.array(possible_colors(ct.dark, BITS_PER_COLOR) + [(255,255,255)])
+        white = state_info['white']
+        ccms = list()
+        i = 0
+        while i < splits:
+            observed = numpy.array([v for k,v in sorted(color_lookups[i].items())] + [white])
+            der = matrix_colour_correction_Cheung2004(observed, exp)
+            ccms.append(der)
+            i += 1
+
+        if ct.ccm is None or cc_setting == 7:
+            ct.ccm = ccms
+        else:
+            ct.ccm = [der.dot(ct.ccm) for der in ccms]
+
+    if cc_setting == 5:
+        ct.colors = color_lookups[0]
+    if cc_setting == 10:
+        ct.disable_color_scaling = True
+        ct.colors = color_lookups[0]
+
+
+def _decode_iter(ct, img, color_img, state_info={}):
+    decoding = sorted(_decode_symbols(ct, img))
+    if use_split_mode():
+        for i, bits, _ in decoding:
+            yield i, bits
+        yield -1, None
+
+    # state_info can be set at any time, but it will probably be set by the caller *after* the empty yield above
+    if state_info.get('color_correct') == 1:
```

Comment:
When this got refactored, the kmeans code got left behind. It should probably make a comeback, because (1) it works, (2) it doesn't require a fountain decode to work, (3) it'll be useful as a comparison.